allow multiple anki note loadouts

StefanVukovic99 opened the pull request:
closes #893

StefanVukovic99 commented:
This is ready for review. Export your settings before checking out the branch!

Kuuuube commented:
A few comments on this from a quick look (havent checked the code):
1. The delete button for note types (card types? notes?) needs to be clearer what it's actually deleting. And it looks a bit out of place how it's put in there at the bottom. If I needed to delete one I would not think of scrolling to the bottom of the dialog as the first thing. Maybe it should be on top or maybe it should instead be on the sticky bottom bar with the close and help button and just detect whichever note type is active.
2. The wording of calling them "Notes" seems weird. Since the main setting is called "Configure Anki card format..." (which should probably be made plural now) maybe we should call these "Anki card formats".
3. The "header" portion of the settings modal takes up a lot of space on the small modal. Maybe it should be forced (or defaulted) to a big modal similar to the "Configure Anki card templates..." setting. It would also help for the distinction between the header settings and everything below it to be clearer. It's not immediately obvious at a glance where one ends and the other starts.

StefanVukovic99 commented:
I've moved the delete button, added an <hr>, and made the modal fullscreen.

I agree we should choose a consistent wording for the notes/cards, at least for the user facing strings. Currently we have "Generate anki notes", "View anki notes", which is technically correct, but in most places we use "cards", which is not correct from Anki's perspective but I think most people use it. 

This looks to be the only place we have tabs in the interface, I think the plus button fits there better than it would in the footer. It's a bit more conspicuous now that I removed the fullscreen button, but maybe it could also get a border or an accent color or something.

jamesmaa commented:
Some notes

<img width="706" alt="image" src="https://github.com/user-attachments/assets/4b7ac21d-c6c1-4356-9cac-081ae4515e32" />
Can we add a hardcoded limit for how many new note types we can add? Something reasonable like 5. Right now it's unbounded which can expose some weird bugs like UI (there are 25 note types but only 18 are displayed). We can always remove limits down the road but if we ship this and want to add limits later, it's going to be more painful

<img width="1075" alt="image" src="https://github.com/user-attachments/assets/166d4579-cc53-458e-a650-b212ec5dded7" />
I noticed that Kanji note type is displayed for non-ja languages. Is it possible to remove this in non-ja languages and default to one note type? 

<img width="372" alt="image" src="https://github.com/user-attachments/assets/20cb0bf6-5b48-486f-a0b5-32d891333347" />
can we have all the icons line downwards for consistency?

jamesmaa reviewed (commented):
Super cool PR! So excited to ship this 🙌

Kuuuube commented:
> Currently we have "Generate anki notes", "View anki notes", which is technically correct, but in most places we use "cards", which is not correct from Anki's perspective but I think most people use it.

My problem here is moreso that these arent cards or notes. When you click the delete button it asks if you want to delete a note. Except these aren't notes and the button doesn't delete any notes. This is an internal Yomitan specific thing which is why I suggest "Anki card formats" instead of "notes".

I could also see thing getting a bit confusing when looking at the code and we have notes that arent actually notes because they have nothing to do directly with Anki's notes. They help to create Anki notes when the user desires, but are not notes themselves.

Kuuuube commented on ext/js/pages/settings/keyboard-shortcuts-controller.js:
```diff
@@ -61,10 +61,9 @@ export class KeyboardShortcutController {
-            ['addNoteTermKanji',                 {scopes: new Set(['popup', 'search'])}],
-            ['addNoteTermKana',                  {scopes: new Set(['popup', 'search'])}],
-            ['viewNotes',                        {scopes: new Set(['popup', 'search'])}],
+            ['addNote1',                         {scopes: new Set(['popup', 'search'])}],
+            ['addNote2',                         {scopes: new Set(['popup', 'search'])}],
+            ['viewNotes1',                       {scopes: new Set(['popup', 'search'])}],
```
Maybe it would be better to just have `addNote` and `viewNotes`  and the user specifies the number. Possibly with a dropdown similar to the `Play audio from source` hotkey. The user will also be able to see the names of their card formats when selecting them this way instead of having to check which one is 1 and 2.

StefanVukovic99 replied:
I didn't know about `Play audio from source`, this definitely looks like the way to go.

StefanVukovic99 replied:
This should work now as you described it. The only downside is the view notes button no longer works on kanji notes.

Kuuuube commented:
This connection line is off center and doesnt fully connect when a small button is selected: 
![image](https://github.com/user-attachments/assets/56fa2bc8-bf5f-46e9-8a0c-7fabadd26f1f)

YukiNagat0 commented:
Repost from discord: https://discord.com/channels/1171709317343346789/1276641729688240229/1360712158945743011
LGTM, the only suggestion is to hide the connection line between main `View added note` button and main `Add` button when there is only one card format.

Kuuuube commented:
> hide the connection line between main `View added note` button and main `Add` button when there is only one card format.

If this is desirable, the `:only-child` css selector can cleanly make this happen:
```css
.note-actions-container .action-button-container:only-child::after {
    display: none;
}
```

StefanVukovic99 commented:
> Repost from discord: https://discord.com/channels/1171709317343346789/1276641729688240229/1360712158945743011 LGTM, the only suggestion is to hide the connection line between main `View added note` button and main `Add` button when there is only one card format.

Yeah, it works that way already

StefanVukovic99 commented:
> Can we add a hardcoded limit for how many new note types we can add? Something reasonable like 5

I added a basic check for this now.

> Is it possible to remove this in non-ja languages and default to one note type? 

I've now set it to create kanji the kanji card format only for `['ja', 'zh', 'yue']`. I don't think anyone's been using kanji dictionaries for korean or viet so that should be fine?

> can we have all the icons line downwards for consistency?

I would keep the first icon group horizontal, because I think lots of people will complain if their view note button gets moved, and if someone has the icons for flags and tags enabled, the flags and tags are probably on the first/main card format and the UI is better when these are shown horizontally. The line should help with any possible confusion at first.

StefanVukovic99 commented on ext/js/data/anki-note-builder.js:
```diff
@@ -46,7 +46,7 @@ export class AnkiNoteBuilder {
+        noteOptions,
         context,
         template,
         deckName,
```
You're right, I think I've removed the redundant stuff now

StefanVukovic99 commented:
> connection line is off center 

applied that css

jamesmaa reviewed (commented):
Poked around a bit on desktop and couldn't find anything broken. Maybe @Kuuuube can poke around as well and give a review?

StefanVukovic99 commented:
@jamesmaa do you have any takes about the naming here? `Notes`, `Card formats`, `Note presets` ... ?

Also, maybe we should rename the card templates setting to "anki field templates" or something to make it more different from however we end up naming this option:
![image](https://github.com/user-attachments/assets/b3512883-b1cb-44bc-91f7-fe512fc7c149)

jamesmaa commented:
> @jamesmaa do you have any takes about the naming here? `Notes`, `Card formats`, `Note presets` ... ?
> 
> Also, maybe we should rename the card templates setting to "anki field templates" or something to make it more different from however we end up naming this option: ![image](https://private-user-images.githubusercontent.com/24891609/435131757-b3512883-b1cb-44bc-91f7-fe512fc7c149.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUwMzkxNDAsIm5iZiI6MTc0NTAzODg0MCwicGF0aCI6Ii8yNDg5MTYwOS80MzUxMzE3NTctYjM1MTI4ODMtYjFjYi00NGJjLTkxZjctZmU1MTJmYzdjMTQ5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE5VDA1MDA0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE5NzQ3N2YwNGQ0MmE3Njc0OWI0NmM5OTQyMDU5YmY4ZWUxNzBlNGUxYmNjZjA4MzczY2EyOTk3M2ZkMmIzZmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.m8YGaZPNGmN63a2tkwHWaJomMLusZT-9laKULiPN5Rs)

If we're trying to avoid the Card vs Note anki terminology, maybe we can write "Configure Anki flashcard fields" and "Configure Anki flashcard handlebar templates"? For non-anki users, "flashcard" is easy to understand what's going on. For power anki users who already know "card" and "note", "flashcard" would be new nomenclature that shouldn't overload any existing terms

jamesmaa commented:
Honestly "Anki card templates" would be a better name for "Anki card format" but we're already using that

Kuuuube commented:
The functionality all looks good for merging this from my testing. Code looks fine.

I'm still not loving calling the card templates (or whatever we should call them) "notes". Also I think the delete button should say more than just "Delete". Maybe "Delete Card Template" (and the confirm dialog should match).

Also though I wouldn't consider it a requirement for this pr, I'm 100% sure someone is going to ask for:
1. An option to make the book button view all card templates at once
2. A way to re-arrange the card formats

As for the "formats" "templates" problem, I think we should just rename the current "templates" option. They can just be named "Anki Handlebars" and they're already called that by many people.
